gematria/datasets/annotating_importer.cc

```diff
@@ -0,0 +1,340 @@
+            importer_.DisassembledInstructionsFromMachineCode(
+                machine_code, begin_idx - file_offset);
+        if (!instructions.ok()) {
+          return instructions.status();
+        }
+        // TODO(vbshah): Consider removing the addresses and just using the
+        // `addr` fields on the instructions.
+        basic_blocks.emplace_back(begin_idx - file_offset,
+                                  end_idx - file_offset, instructions.value());
+      }
+    }
+  }
+
+  return basic_blocks;
+}
+
+absl::StatusOr<std::pair<std::vector<std::string>,
+                         std::unordered_map<uint64_t, std::vector<int>>>>
+AnnotatingImporter::GetSamples() {
+  const quipper::PerfDataProto &perf_data_proto = perf_reader_.proto();
+
+  // Extract event type information.
+  std::vector<std::string> sample_types(perf_data_proto.event_types_size());
+  std::unordered_map<int, int> event_code_to_idx;
+  for (int sample_type_idx = 0; sample_type_idx < sample_types.size();
+       ++sample_type_idx) {
+    const auto &event_type = perf_data_proto.event_types()[sample_type_idx];
+    sample_types[sample_type_idx] = event_type.name();
+    event_code_to_idx[event_type.id()] = sample_type_idx;
+  }
+  std::unordered_map<int, int> event_id_to_code;
+  for (const auto &event_type : perf_data_proto.file_attrs()) {
+    // Mask out bits identifying the PMU and not the event.
+    int event_code = event_type.attr().config() & 0xffff;
+    for (int event_id : event_type.ids()) {
+      event_id_to_code[event_id] = event_code;
+    }
+  }
+
+  // Find the relevant mapping.
+  quipper::PerfDataProto::MMapEvent mmap;
+  for (const auto &event : perf_data_proto.events()) {
+    // TODO(virajbshah): Not sure if this always works, i.e. does the main
+    // binary always correspond to the first MMapEvent. Implement BuildID or
+    // name based checking.
+    if (event.has_mmap_event() &&
+        event.mmap_event().prot() & 1 /* PROT_READ */ &&
+        event.mmap_event().prot() & 4 /* PROT_EXEC */) {
+      mmap = event.mmap_event();
+      break;
+    }
+  }
+  // TODO(virajbshah): Make sure the mapping was found. (Use num_mmap_events)
+
+  // If the profile has multiple event types, lookups are needed to find the
+  // event type corresponding to a sample. In the other case, this is neither
+  // required nor possible - since samples are not associated with IDs for
+  // lookup in the proto.
+  const bool has_multiple_sample_types = sample_types.size() > 1;
+
+  // Process sample events.
+  std::unordered_map<uint64_t, std::vector<int>> samples;
+  for (const auto &event : perf_data_proto.events()) {
+    // Filter out non-sample events.
+    if (!event.has_sample_event()) {
+      continue;
+    }
+
+    // Filter out sample events from outside the profiled binary.
+    uint64_t sample_ip = event.sample_event().ip();
+    if (!(mmap.start() <= sample_ip && sample_ip < mmap.start() + mmap.len())) {
```

Comment:
1. Simpler, with De Morgan's laws:
```
if (sample_ip < mmap_start || sample_id >= mmap_end) continue;
```

2. `mmap.start() + mmap.end()` is always the same, since we track just one `mmap` event, so we could pre-compute it for shorter and more efficient code.